filters/auth/oauth.go

```diff
@@ -0,0 +1,343 @@
+	groupServiceAccess rejectReason = "group-service-access"
+	invalidGroup       rejectReason = "invalid-group"
+)
+
+const (
+	AuthName      = "auth"
+	AuthGroupName = "authGroup"
+	// BasicAuthName = "basicAuth"
+)
+
+type (
+	authClient  struct{ urlBase string }
+	groupClient struct{ urlBase string }
+
+	authDoc struct {
+		Uid    string   `json:"uid"`
+		Realm  string   `json:"realm"`
+		Scopes []string `json:"scope"` // TODO: verify this with service2service authentication
+	}
+
+	groupDoc struct {
+		Id string `json:"id"`
+	}
+
+	spec struct {
+		typ        roleCheckType
+		authClient *authClient
+	}
+
+	filter struct {
+		typ         roleCheckType
+		authClient  *authClient
+		groupClient *groupClient
+		realm       string
+		args        []string
+	}
+
+	// basic string
+)
+
+var (
+	errInvalidAuthorizationHeader = errors.New("invalid authorization header")
+	errInvalidToken               = errors.New("invalid token")
+)
+
+func getToken(r *http.Request) (string, error) {
+	const b = "Bearer "
+	h := r.Header.Get(authHeaderName)
+	if !strings.HasPrefix(h, b) {
+		return "", errInvalidAuthorizationHeader
+	}
+
+	return h[len(b):], nil
+}
+
+func unauthorized(ctx filters.FilterContext, uname string, reason rejectReason) {
+	ctx.StateBag()[logfilter.AuthUserKey] = uname
+	ctx.StateBag()[logfilter.AuthRejectReasonKey] = string(reason)
+	ctx.Serve(&http.Response{StatusCode: http.StatusUnauthorized})
+}
+
+func authorized(ctx filters.FilterContext, uname string) {
+	ctx.StateBag()["auth-user"] = uname
+}
+
+func getStrings(args []interface{}) ([]string, error) {
+	s := make([]string, len(args))
+	var ok bool
+	for i, a := range args {
+		s[i], ok = a.(string)
+		if !ok {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return s, nil
+}
+
+func intersect(left, right []string) bool {
+	for _, l := range left {
+		for _, r := range right {
+			if l == r {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func jsonGet(url, auth string, doc interface{}) error {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	if auth != "" {
+		req.Header.Set(authHeaderName, "Bearer "+auth)
+	}
+
+	rsp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer rsp.Body.Close()
+	if rsp.StatusCode != 200 {
+		return errInvalidToken
+	}
+
+	d := json.NewDecoder(rsp.Body)
+	return d.Decode(doc)
+}
+
+func (ac *authClient) validate(token string) (*authDoc, error) {
+	var a authDoc
+	err := jsonGet(ac.urlBase, token, &a)
+	return &a, err
+}
+
+func (tc *groupClient) getGroups(uid, token string) ([]string, error) {
+	var t []groupDoc
+	err := jsonGet(tc.urlBase+uid, token, &t)
+	if err != nil {
+		return nil, err
+	}
+
+	ts := make([]string, len(t))
+	for i, ti := range t {
+		ts[i] = ti.Id
+	}
+
+	return ts, nil
+}
+
+func newSpec(typ roleCheckType) filters.Spec {
+	return &spec{typ: typ}
+}
+
+// Creates a new auth filter specification to validate authorization
```

Comment:
godoc forces you to start with function name, so:
"NewAuth creates a new filter specification..."